astropy/units/quantity.py

```diff
@@ -629,13 +638,17 @@ def _set_unit(self, unit):
         a unit is consistent.
         """
         if not isinstance(unit, UnitBase):
-            # Trying to go through a string ensures that, e.g., Magnitudes with
-            # dimensionless physical unit become Quantity with units of mag.
-            unit = Unit(str(unit), parse_strict='silent')
-            if not isinstance(unit, UnitBase):
-                raise UnitTypeError(
-                    "{} instances require {} units, not {} instances."
-                    .format(type(self).__name__, UnitBase, type(unit)))
+            if (isinstance(self._unit, StructuredUnit)
```

Comment:
Relates to concern below about ``StructuredUnit`` not inheriting from UnitBase.

Reply:
Let's discuss on the main thread.